src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -639,6 +639,119 @@ void SpatialPooler::stripUnlearnedColumns(UInt activeArray[]) const
   }
 }
 
+bool SpatialPooler::operator==(const SpatialPooler &sp) const
+{
+    const UInt numColumns = this->getNumColumns();
+    const UInt numInputs = this->getNumInputs();
+
+    ASSERT_TRUE(this->getNumColumns() == sp.getNumColumns());
+    ASSERT_TRUE(this->getNumInputs() == sp.getNumInputs());
```

Comment:
RFC: I'm looking for a way how to approach implementation of these comparisons.

The method in SPTest class used manual  `ASSERT_TRUE()` calls, the advantage is that when there is an inequality you'll know what exactly caused it.

Now for `equals==` implementation we need to turn the asserts into a conjunction of conditions `(a1==b1 && a2==b2 &&...)`.

I was thinking of turning this
```
    ASSERT_TRUE(this->getNumColumns() == sp.getNumColumns());
    ASSERT_TRUE(this->getNumInputs() == sp.getNumInputs());
```
Into

   * A) `(a1==b1) || return false` ; but this result in losing the information on which specific check the comparison failed. Is that a problem?
   * B) If the above is a problem, we could return a huge conjunction
    `return (this->getNumColumns() == sp.getNumColumns()) && (this->getNumInputs()==sp.getNumInputs()) && ..`

I think the b) would inform us in gtest at which part the conjunction failed, keeping the desired behavior.
The only problem with this is the readability and that some checks have pre-processing or even are inside for-loops, but we could workaround that.

    * C) keep the asserts (`EXPECT_EQ`) for information and just add `&&` for equals.

EDIT:
add option C)

Reply:
Update: I'll keep the asserts, resp NTA_CHECK, which will provide the information. And wrap in try-catch and return false if throws